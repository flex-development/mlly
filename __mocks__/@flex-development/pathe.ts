/**
 * @file Mocks - @flex-development/pathe
 * @module mocks/flex-development/pathe
 * @see https://github.com/flex-development/pathe
 */

/**
 * [`@flex-development/pathe`][1] module type.
 *
 * [1]: https://github.com/flex-development/pathe
 */
type Actual = typeof import('@flex-development/pathe')

/**
 * `@flex-development/pathe` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/pathe')

export default {
  basename: vi.fn(actual.basename),
  changeExt: vi.fn(actual.changeExt),
  dirname: vi.fn(actual.dirname),
  extname: vi.fn(actual.extname),
  format: vi.fn(actual.format),
  isAbsolute: vi.fn(actual.isAbsolute),
  join: vi.fn(actual.join),
  parse: vi.fn(actual.parse),
  relative: vi.fn(actual.relative),
  removeExt: vi.fn(actual.removeExt),
  resolve: vi.fn(actual.resolve)
}
