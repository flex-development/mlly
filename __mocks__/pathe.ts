/**
 * @file Mocks - pathe
 * @module mocks/pathe
 * @see https://github.com/unjs/pathe
 */

/**
 * [`pathe`][1] module type.
 *
 * [1]: https://github.com/unjs/pathe
 */
type Actual = typeof import('pathe')

/**
 * `pathe` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('pathe')

export const basename = vi.fn(actual.basename)
export const dirname = vi.fn(actual.dirname)
export const extname = vi.fn(actual.extname)
export const isAbsolute = vi.fn(actual.isAbsolute)
export const join = vi.fn(actual.join)
export const parse = vi.fn(actual.parse)
export const relative = vi.fn(actual.relative)
export const resolve = vi.fn(actual.resolve)
