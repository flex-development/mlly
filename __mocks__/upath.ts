/**
 * @file Mocks - upath
 * @module mocks/upath
 * @see https://github.com/anodynos/upath
 */

/**
 * [`upath`][1] module type.
 *
 * [1]: https://github.com/anodynos/upath
 */
type Actual = typeof import('upath')

/**
 * `upath` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('upath')

export default {
  basename: vi.fn(actual.basename),
  dirname: vi.fn(actual.dirname),
  extname: vi.fn(actual.extname),
  format: vi.fn(actual.format),
  isAbsolute: vi.fn(actual.isAbsolute),
  join: vi.fn(actual.join),
  parse: vi.fn(actual.parse),
  relative: vi.fn(actual.relative),
  resolve: vi.fn(actual.resolve)
}
