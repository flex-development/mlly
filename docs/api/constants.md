---
head:
  - - meta
    - name: description
      content: Constant values exported from @flex-development/mlly
  - - meta
    - name: keywords
      content:
        - ecmascript
        - es modules
        - esm
        - node
  - - meta
    - name: priority
      content: 0.8
---

# Constants

Constant values exported from `mlly`.

These values are used as defaults for several [library functions](index.md).

<script setup lang="ts">
import { useData } from 'vitepress'
import type ThemeConfig from '../.vitepress/theme/config'
import type Documentation from '../.vitepress/theme/documentation'

const { site } = useData<ThemeConfig>()
const { documentation } = site.value.themeConfig

/**
 * Documentation objects.
 *
 * @const {Documentation[]} docs
 */
const docs: Documentation[] = documentation.filter(doc => {
  return /src\/constants\.ts$/.test(doc.file)
})
</script>

<Doc v-for="doc in docs" :doc="doc.doc" :key="doc.file" />
