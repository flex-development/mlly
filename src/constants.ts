/**
 * @file Constants
 * @module mlly/constants
 */

/**
 * CommonJS syntax regex.
 *
 * @const {RegExp} CJS_SYNTAX_REGEX
 */
export const CJS_SYNTAX_REGEX: RegExp =
  /(?<!(?:\/\/|\*).*)((?:module\.)?exports(?:\.\w+|(?<!\s+=))|require(?=\(.)|require\.\w+|__dirname|__filename)/gm

/**
 * ESM syntax regex.
 *
 * @const {RegExp} ESM_SYNTAX_REGEX
 */
export const ESM_SYNTAX_REGEX: RegExp =
  // eslint-disable-next-line unicorn/no-unsafe-regex
  /(?<!(?:\/\/|\*).*)((?:export|import)[\s\w*,{}]*(?=\sfrom)|export\b\s*(?:[*{]|async function|(?:abstract\s)?class|const|default|enum|function|interface|let|type|var)|await import|import\.meta\.(?:env(?:\.\w+)?|resolve|url))/gm
