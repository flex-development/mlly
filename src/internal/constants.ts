/**
 * @file Internals - Constants
 * @module mlly/internal/constants
 */

/**
 * CommonJS syntax regex.
 *
 * @internal
 *
 * @const {RegExp} CJS_SYNTAX_REGEX
 */
export const CJS_SYNTAX_REGEX: RegExp =
  /(?<!(?:\/\/|\*).*)((?:module\.)?exports(?:\.\w+|(?<!\s+=))|require(?=\(.)|require\.\w+|__dirname|__filename|await import)/gm

/**
 * Dynamic import statement regex.
 *
 * @internal
 *
 * @const {RegExp} DYNAMIC_IMPORT_REGEX
 */
export const DYNAMIC_IMPORT_REGEX: RegExp =
  /(?<!(?:\/\/|\*).*)(?:(?:const|let|var)(?:(?<name>\s+\w+\s*)|\s*{\s*(?<names>\w+(?:,\s*\w+)*)\s*}\s*)(?:\s*=\s*?)?.)?(?:await)?\s+import\s*\((?<expression>(?:[^()]+|\((?:[^()]+|\([^()]*\))*\))*)\)/gm

/**
 * ESM syntax regex.
 *
 * @internal
 *
 * @const {RegExp} ESM_SYNTAX_REGEX
 */
export const ESM_SYNTAX_REGEX: RegExp =
  /(?<!(?:\/\/|\*).*)((?:export|import)[\s\w*,{}]*(?=\sfrom)|export\b\s*(?:[*{]|async function|(?:abstract\s)?class|const|default|enum|function|interface|let|type|var)|await import|import\.meta\.(?:env(?:\.\w+)?|resolve|url))/gm

/**
 * Declaration export statement regex.
 *
 * **Note**: Captures [declaration type][1] and export name only.
 *
 * [1]: ./types/declaration.ts
 *
 * @const {RegExp} EXPORT_DECLARATION_REGEX
 */
export const EXPORT_DECLARATION_REGEX: RegExp =
  /(?<!(?:\/\/|\*).*)\bexport(?: +declare)? +(?<declaration>(?:(?:abstract +)?class|(?:default +)?async +function\*?|(?:default +)?function\*?|default +async|default(?!(?: +async)? +function\*?)|(?:const +)?enum|const|interface|let|namespace|type(?! *{)|var)) +(?<name>[\w$]+(?!.*\b +=>))?/gm

/**
 * Named export statement regex.
 *
 * @internal
 *
 * @const {RegExp} EXPORT_NAMED_REGEX
 */
export const EXPORT_NAMED_REGEX: RegExp =
  /(?<!(?:\/\/|\*).*)\bexport(?:\s+type)?\s+{(?<exports>[^}]+?)[\s,]*}(?:\s*from\s*["']\s*(?<specifier>(?<="\s*)[^"]*[^\s"](?=\s*")|(?<='\s*)[^']*[^\s'](?=\s*'))\s*["'][^\n;]*)?/gm

/**
 * Aggregate export statement (`export * ...`) regex.
 *
 * @internal
 *
 * @const {RegExp} EXPORT_STAR_REGEX
 */
export const EXPORT_STAR_REGEX: RegExp =
  /(?<!(?:\/\/|\*).*)\bexport\s*\*(?:\s*as\s+(?<name>[\w$]+)\s+)?\s*(?:\s*from\s*["']\s*(?<specifier>(?<="\s*)[^"]*[^\s"](?=\s*")|(?<='\s*)[^']*[^\s'](?=\s*'))\s*["'][^\n;]*)?/gm

/**
 * Require statement regex.
 *
 * @internal
 *
 * @const {RegExp} REQUIRE_STATEMENT_REGEX
 */
export const REQUIRE_STATEMENT_REGEX: RegExp =
  /(?<!(?:\/\/|\*).*)(?:\bconst[ {]+(?<imports>[\w\t\n\r $*,/]+)[ =}]+)?(?<type>\brequire(?:\.resolve)?)\(["'] *(?<specifier>(?<="\s*)[^"']*[^\s"'](?=\s*")|(?<=["']\s*)[^']*[^\s'](?=\s*')) *["']\)/gm

/**
 * Static import statement regex.
 *
 * @internal
 *
 * @const {RegExp} STATIC_IMPORT_REGEX
 */
export const STATIC_IMPORT_REGEX: RegExp =
  /(?<!(?:\/\/|\*).*)import\s*(?:[\s"']*(?<imports>[\w\t\n\r $*,/{}]+)from\s*)?["']\s*(?<specifier>(?<="\s*)[^"]*[^\s":](?=\s*")|(?<=["']\s*)[^']*[^\s':](?=\s*'))\s*["']\s*/gm
