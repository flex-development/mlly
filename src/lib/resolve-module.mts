/**
 * @file resolveModule
 * @module mlly/lib/resolveModule
 */

import defaultExtensions from '#lib/default-extensions'
import resolveAlias from '#lib/resolve-alias'
import { moduleResolve } from '#lib/resolver'
import {
  codes,
  isNodeError,
  type Code,
  type ErrModuleNotFound,
  type NodeError
} from '@flex-development/errnode'
import type {
  ChangeExtFn,
  ModuleId,
  ResolveModuleOptions
} from '@flex-development/mlly'
import pathe from '@flex-development/pathe'

export default resolveModule

/**
 * Resolve a module `specifier` according to the [ESM Resolver algorithm][esm],
 * mostly ðŸ˜‰.
 *
 * Adds support for:
 *
 * - Changing file extensions
 * - Directory index resolution
 * - Extensionless file resolution
 * - Path alias resolution
 * - Scopeless `@types/*` resolution (i.e. `unist` -> `@types/unist`)
 *
 * [esm]: https://nodejs.org/api/esm.html#esm_resolver_algorithm
 *
 * @see {@linkcode ModuleId}
 * @see {@linkcode NodeError}
 * @see {@linkcode ResolveModuleOptions}
 *
 * @param {string} specifier
 *  The module specifier to resolve
 * @param {ModuleId} parent
 *  URL of parent module
 * @param {ResolveModuleOptions | null | undefined} [options]
 *  Resolution options
 * @return {URL}
 *  Resolved URL
 * @throws {NodeError}
 */
function resolveModule(
  specifier: string,
  parent: ModuleId,
  options?: ResolveModuleOptions | null | undefined
): URL {
  try {
    return changeExt(moduleResolve(
      resolveAlias(specifier, { ...options, parent }) ?? specifier,
      parent,
      options?.conditions,
      options?.mainFields,
      options?.preserveSymlinks,
      options?.fs
    ), specifier, options?.ext)
  } catch (e: unknown) {
    /**
     * Error codes to ignore when attempting to resolve {@linkcode specifier}.
     *
     * @const {Set<Code>} ignore
     */
    const ignore: Set<Code> = new Set<Code>([
      codes.ERR_MODULE_NOT_FOUND,
      codes.ERR_UNSUPPORTED_DIR_IMPORT
    ])

    if (isNodeError(e) && ignore.has(e.code)) {
      /**
       * Module extensions to probe for.
       *
       * @const {string[]} extensions
       */
      const extensions: string[] = [
        ...(options?.extensions ?? defaultExtensions)
      ]

      /**
       * Module specifiers to try resolving.
       *
       * @var {string[]} tries
       */
      let tries: string[] = []

      // add @types resolution attempts if package resolution failed.
      if (
        e.code === codes.ERR_MODULE_NOT_FOUND &&
        !(e as ErrModuleNotFound).url
      ) {
        tries = [
          specifier.startsWith('@types/') ? specifier : '@types/' + specifier
        ].flatMap(specifier => [
          specifier,
          specifier + pathe.sep + 'index.d.ts',
          specifier + pathe.sep + 'index.d.mts',
          specifier + pathe.sep + 'index.d.cts'
        ])
      }

      // add extensionless file resolution attempts if file resolution failed.
      if (
        e.code === codes.ERR_MODULE_NOT_FOUND &&
        (e as ErrModuleNotFound).url
      ) {
        tries = extensions.map(ext => specifier + pathe.formatExt(ext))
      }

      // add directory index resolution attempts if directory resolution failed.
      if (e.code === codes.ERR_UNSUPPORTED_DIR_IMPORT) {
        tries = extensions.map(ext => {
          return specifier + pathe.sep + 'index' + pathe.formatExt(ext)
        })
      }

      // try module resolution attempts.
      for (const attempt of tries) {
        try {
          return changeExt(moduleResolve(
            attempt,
            parent,
            options?.conditions,
            options?.mainFields,
            options?.preserveSymlinks,
            options?.fs
          ), specifier, options?.ext)
        } catch {
          // swallow error to continue resolution attempts.
          continue
        }
      }
    }

    throw e
  }
}

/**
 * Change the file extension of `url`.
 *
 * @internal
 *
 * @param {URL} url
 *  The resolved module URL
 * @param {string} specifier
 *  The module specifier being resolved
 * @param {ChangeExtFn | string | null | undefined} [ext]
 *  Replacement file extension or function that returns a file extension
 * @return {URL}
 *  `url`
 */
function changeExt(
  url: URL,
  specifier: string,
  ext?: ChangeExtFn | string | null | undefined
): URL {
  if (url.protocol === 'file:' && ext !== undefined) {
    url.href = typeof ext === 'function'
      ? pathe.changeExt(url.href, ext(url, specifier))
      : pathe.changeExt(url.href, ext)

    url = new URL(pathe.toPosix(url.href).replace(/\/index$/, pathe.sep))
  }

  return url
}
