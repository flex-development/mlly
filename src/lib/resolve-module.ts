/**
 * @file resolveModule
 * @module mlly/lib/resolveModule
 */

import { CONDITIONS, RESOLVE_EXTENSIONS } from '#src/constants'
import type { ResolveOptions as Options } from '#src/interfaces'
import type { SpecifierType } from '#src/types'
import isBuiltin from '@flex-development/is-builtin'
import { moduleResolve, type ErrnoException } from 'import-meta-resolve'
import { pathToFileURL, URL } from 'node:url'
import upath from 'upath'
import toBareSpecifier from './to-bare-specifier'
import toRelativeSpecifier from './to-relative-specifier'

/**
 * Resolves `specifier` according to the [ESM Resolver algorithm][1], mostly ðŸ˜‰.
 *
 * Adds support for:
 *
 * - Resolving without file extensions and explicit `/index` usage
 * - Resolving `@types/*` with **and** without explicit `@types/*` usage
 * - Converting resolved modules into bare and relative specifiers
 * - Removing and replacing file extensions
 *
 * [1]: https://nodejs.org/api/esm.html#esm_resolver_algorithm
 * [2]: ../interfaces/resolve-options.ts
 * [3]: ./to-bare-specifier.ts
 * [4]: ./to-relative-specifier.ts
 *
 * @see [`Options`][2]
 * @see [`toBareSpecifier`][3]
 * @see [`toRelativeSpecifier`][4]
 *
 * @async
 *
 * @param {string} specifier - Module specifier to resolve
 * @param {Options} [options={}] - Resolve options
 * @return {Promise<string>} Resolved module
 * @throws {ErrnoException}
 */
const resolveModule = async (
  specifier: string,
  options: Options = {}
): Promise<string> => {
  const {
    conditions = CONDITIONS,
    ext,
    extensions = RESOLVE_EXTENSIONS,
    parent = import.meta.url,
    preserveSymlinks = false,
    ...opts
  } = options

  /**
   * Parent module URL to resolve {@link specifier} from.
   *
   * @const {URL} base
   */
  const base: URL =
    typeof parent === 'string'
      ? parent.startsWith('file:')
        ? new URL(parent)
        : pathToFileURL(parent)
      : parent

  /**
   * Module specifiers to try resolving.
   *
   * @const {string[]} tries
   */
  const tries: string[] =
    isBuiltin(specifier) || /^(?:data|https?):/.test(specifier)
      ? []
      : extensions.flatMap(ext => {
          return [specifier + ext, specifier + '/index' + ext].flatMap($try => {
            return [$try, '@types/' + $try]
          })
        })

  // ensure attempt to resolve original specifier is first
  tries.unshift(specifier)

  /**
   * {@link moduleResolve} error codes to ignore when attempting to resolve
   * module ids without URL schemes.
   *
   * **Note**: If an error is thrown, it'll be reported **_after_** all module
   * ids in {@link tries} have been evaluated.
   *
   * @const {Set<NonNullable<ErrnoException['code']>>} ignore
   */
  const ignore: Set<NonNullable<ErrnoException['code']>> = new Set([
    'ERR_MODULE_NOT_FOUND',
    'ERR_PACKAGE_PATH_NOT_EXPORTED',
    'ERR_UNSUPPORTED_DIR_IMPORT'
  ])

  /**
   * Resolution error.
   *
   * @var {ErrnoException | undefined} error
   */
  let error: ErrnoException | undefined

  /**
   * Resolved module specifier.
   *
   * @var {string | undefined} resolved
   */
  let resolved: string | undefined

  // try module resolution
  for (const id of tries) {
    try {
      resolved = moduleResolve(
        id,
        base,
        new Set(conditions),
        preserveSymlinks
      ).href
    } catch (e: unknown) {
      if (id === specifier) error = e as ErrnoException
      if (ignore.has((e as ErrnoException).code!)) continue

      throw e
    }

    if (resolved) break
  }

  // throw error if specifier could not be resolved
  if (!resolved) throw error!

  // return resolved early if extension or type logic should be skipped
  if (isBuiltin(resolved) || /^(?:data|https?):/.test(resolved)) return resolved

  /**
   * Module specifier type.
   *
   * @const {SpecifierType} type
   */
  const type: SpecifierType =
    typeof opts.type === 'function'
      ? await opts.type(resolved)
      : opts.type ?? 'absolute'

  // convert resolved into bare specifier
  if (type === 'bare') resolved = await toBareSpecifier(resolved, conditions)

  // convert resolved into relative specifier
  if (type === 'relative') resolved = toRelativeSpecifier(resolved, parent)

  // remove or replace file extension
  if (ext !== undefined) {
    if (ext === false) {
      resolved = upath.removeExt(resolved, upath.extname(resolved))
    } else if (type === 'relative') {
      resolved = upath.changeExt(
        resolved,
        typeof ext === 'function' ? await ext(specifier, resolved) : ext
      )
    }

    resolved = resolved.replace(/\/index$/, '')
  }

  return resolved
}

export default resolveModule
