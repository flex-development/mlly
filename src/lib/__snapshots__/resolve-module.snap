// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:lib/resolveModule > should return resolved URL ("#internal/fs") 1`] = `"node:fs/promises"`;

exports[`unit:lib/resolveModule > should return resolved URL ("../../../src") 1`] = `file://\${process.cwd()}/src/index.mts`;

exports[`unit:lib/resolveModule > should return resolved URL ("../../lib") 1`] = `file://\${process.cwd()}/src/lib/index.mjs`;

exports[`unit:lib/resolveModule > should return resolved URL ("../resolve-module") 1`] = `file://\${process.cwd()}/src/lib/resolve-module.mjs`;

exports[`unit:lib/resolveModule > should return resolved URL ("@flex-development/errnode") 1`] = `file://\${process.cwd()}/node_modules/@flex-development/errnode/dist/index.mjs`;

exports[`unit:lib/resolveModule > should return resolved URL ("@flex-development/mlly/package.json") 1`] = `file://\${process.cwd()}/package.json`;

exports[`unit:lib/resolveModule > should return resolved URL ("@types/unist") 1`] = `file://\${process.cwd()}/node_modules/@types/unist/index.d.ts`;

exports[`unit:lib/resolveModule > should return resolved URL ("data:text/javascript,") 1`] = `"data:text/javascript,"`;

exports[`unit:lib/resolveModule > should return resolved URL ("fs") 1`] = `"node:fs"`;

exports[`unit:lib/resolveModule > should return resolved URL ("https://esm.sh/@flex-development/mlly") 1`] = `"https://esm.sh/@flex-development/mlly"`;

exports[`unit:lib/resolveModule > should return resolved URL ("legacy-main-1") 1`] = `file://\${process.cwd()}/__fixtures__/node_modules/legacy-main-1/index.js`;

exports[`unit:lib/resolveModule > should return resolved URL ("legacy-main-2") 1`] = `file://\${process.cwd()}/__fixtures__/node_modules/legacy-main-2/index.json`;

exports[`unit:lib/resolveModule > should return resolved URL ("node:fs/promises") 1`] = `"node:fs/promises"`;

exports[`unit:lib/resolveModule > should return resolved URL ("unist") 1`] = `file://\${process.cwd()}/node_modules/@types/unist/index.d.ts`;
